File: 72412.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <iostream>
#include <algorithm>

using namespace std;

unordered_map<string, vector<int>> formatted;

void parseGivenInfo(string &info, int pointer, string parsed) {
    string current;
    
    if(info[pointer] >= 48 && info[pointer] <= 57) {
        formatted[parsed].push_back(stoi(info.substr(pointer)));
        return;
    }
        
    string check = info.substr(pointer, 2);
    if(check == "ja") current = "java";
    else if(check == "py") current = "python";
    else if(check == "cp") current = "cpp";
    else if(check == "ba") current = "backend";
    else if(check == "fr") current = "frontend";
    else if(check == "ju") current = "junior";
    else if(check == "se") current = "senior";
    else if(check == "ch") current = "chicken";
    else if(check == "pi") current = "pizza";
    
    parseGivenInfo(info, pointer + current.size() + 1, parsed + current[0]);
    parseGivenInfo(info, pointer + current.size() + 1, parsed + '-');
}

void processQuery(string &query, int pointer, string parsed, vector<int> &answer) {
    string current;
    
    if(query[pointer] >= 48 && query[pointer] <= 57) {
        int score = stoi(query.substr(pointer));
        vector<int> scores = formatted[parsed];

        int count = 0;
        for(auto each : scores) if(each >= score) count++;
        answer.push_back(count);
        return;
    }
    
    string check = query.substr(pointer, 2);
    if(check == "- ") current = "-";
    else if(check == "ja") current = "java";
    else if(check == "py") current = "python";
    else if(check == "cp") current = "cpp";
    else if(check == "ba") current = "backend";
    else if(check == "fr") current = "frontend";
    else if(check == "ju") current = "junior";
    else if(check == "se") current = "senior";
    else if(check == "ch") current = "chicken";
    else if(check == "pi") current = "pizza";
    
    if(query[pointer + current.size() + 1] == 'a') processQuery(query, pointer + current.size() + 5, parsed + current[0], answer);
    else processQuery(query, pointer + current.size() + 1, parsed + current[0], answer);
}

vector<int> solution(vector<string> info, vector<string> query) {
    vector<int> answer;
    for(int i = 0; i < info.size(); i++) parseGivenInfo(info[i], 0, "");
    for(auto each : query) processQuery(each, 0, "", answer);
    
    return answer;
}

File: 72412.swift
import Foundation

func solution(_ info:[String], _ query:[String]) -> [Int] {
    var answer: [Int] = []
    var candidates: [String: [Int]] = [:]

    for each in info.enumerated() {
        var parsed = each.1.components(separatedBy: " ")
        guard let last = parsed.popLast(), let score = Int(last) else { return [] }
        makeUpCandidates(parsed, score, 0, &candidates)
    }
    
    for (query, scores) in candidates { candidates[query] = scores.sorted { $0 < $1 } }
    
    for each in query {
        var parsed = each.components(separatedBy: " ").filter { $0 != "and" }
        guard let last = parsed.popLast(), let score = Int(last) else { return [] }
        let query = parsed.joined()
        if let candidate = candidates[query] {
            answer.append(binarySearch(candidate, score))
        } else {
            answer.append(0)
        }
    }

    return answer
}

func makeUpCandidates(_ parsed: [String], _ score: Int, _ index: Int, _ candidates: inout [String: [Int]]) {
    if index == 4 {
        let joined = parsed.joined()
        if candidates[joined] != nil {
            candidates[joined]?.append(score)
        } else {
            candidates[joined] = [score]
        }
        return
    }
    
    makeUpCandidates(parsed, score, index + 1, &candidates)
    var temp = parsed
    temp[index] = "-"
    makeUpCandidates(temp, score, index + 1, &candidates)
}

func binarySearch(_ candidates: [Int], _ target: Int) -> Int {
    var from = 0
    var to = candidates.count
    var mid = (from + to) / 2
    
    while to > from {
        if candidates[mid] < target {
            from = mid + 1
            mid = (from + to) / 2
        } else {
            to = mid
            mid = (from + to) / 2
        }
    }
    
    return candidates.count - from
}
